language/move-lang/stdlib/modules/fixedpoint32.move

```diff
@@ -0,0 +1,62 @@
+address 0x0:
+
+module FixedPoint32 {
+
+    // Define a fixed-point numeric type with 32 fractional bits.
+    // This is just a u64 integer but it is wrapped in a struct to
+    // make a unique type.
+    struct T {
+        value: u64,
+    }
+
+    // Multiply a u64 integer by a fixed-point number, truncating any
+    // fractional part of the product.
+    public fun multiply_u64(num: u64, multiplier: T): u64 {
+        // The product of two 64 bit values has 128 bits, so perform the
+        // multiplication with u128 types and keep the full 128 bit product
+        // to avoid losing accuracy.
+        let unscaled_product = (num as u128) * (multiplier.value as u128);
+        // The unscaled product has 32 fractional bits (from the multiplier)
+        // so rescale it by shifting away the low bits.
+        let product = unscaled_product >> 32;
+        // If the multiplier is larger than 1.0, this can overflow; check that.
+        if ((product >> 64) > 0) abort 10;
+        (product as u64)
+    }
+
+    // Divide a u64 integer by a fixed-point number, truncating any
+    // fractional part of the quotient.
+    public fun divide_u64(num: u64, divisor: T): u64 {
+        // First convert to 128 bits and then shift left to
+        // add 32 fractional zero bits to the dividend.
+        let scaled_value = (num as u128) << 32;
+        // Divide and convert the quotient to 64 bits.
+        let quotient = scaled_value / (divisor.value as u128);
+        // If the divisor is less than 1.0, this can overflow; check that.
+        if ((quotient >> 64) > 0) abort 10;
+        (quotient as u64)
+    }
+
+    public fun get_raw_value(num: T): u64 {
+        num.value
+    }
+
+    public fun create_from_raw_value(value: u64): T {
+        T { value }
+    }
+
+    // Create a fixed-point value from a rational number specified by its
+    // numerator and denominator. This function is for convenience; it is also
+    // perfectly fine to create a fixed-point value by directly specifying the
+    // raw value.
+    public fun create_from_rational(numerator: u64, denominator: u64): T {
+	// Scale the numerator to have 64 fractional bits and the denominator
+	// to have 32 fractional bits, so that the quotient will have 32
+	// fractional bits.
+	let scaled_numerator = (numerator as u128) << 64;
+	let scaled_denominator = (denominator as u128) << 32;
+	// Divide. This cannot overflow.
+	let quotient = scaled_numerator / scaled_denominator;
+	T { value: (quotient as u64) }
```

Comment:
Does this cast fail if the number is too large?

Reply:
it will in the runtime for what is worth

Reply:
I don't know what I was thinking when I wrote that comment -- probably that these were only 32 bit values. Of course this can overflow. The current behavior of integer casts is that they truncate, but @vgao1996 is planning to change that. In the meantime, I can add an explicit check here.